bot/internal/bot/changelog.go

```diff
@@ -0,0 +1,159 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package bot
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"regexp"
+	"strings"
+	"unicode"
+
+	"github.com/gravitational/trace"
+	"k8s.io/utils/strings/slices"
+)
+
+const NoChangelogLabel string = "no-changelog"
+const ChangelogPrefix string = "changelog: "
+const ChangelogTag string = "changelog"
+
+func (b *Bot) Changelog(ctx context.Context) error {
+	pull, err := b.c.GitHub.GetPullRequest(ctx,
+		b.c.Environment.Organization,
+		b.c.Environment.Repository,
+		b.c.Environment.Number,
+	)
+	if err != nil {
+		return trace.Wrap(err, "failed to retrieve pull request for https://github.com/%s/%s/pull/%d", b.c.Environment.Organization, b.c.Environment.Repository, b.c.Environment.Number)
+	}
+
+	if slices.Contains(pull.UnsafeLabels, NoChangelogLabel) {
+		log.Printf("PR contains no changelog label %q, skipping changelog check", NoChangelogLabel)
+		return nil
+	}
+
+	changelogEntry, err := b.getChangelogEntry(ctx, pull.UnsafeBody)
+	if err != nil {
+		return trace.Wrap(err, "failed to get changelog entry")
+	}
+
+	err = b.validateChangelogEntry(ctx, changelogEntry)
+	if err != nil {
+		return trace.Wrap(err, "failed to validate changelog entry")
+	}
+
+	err = b.deleteComments(ctx)
```

Comment:
TBH I would consider simplifying significantly, just check the PR description and leave a comment if changelog is missing or incorrect, without trying to update or delete existing comments. Wouldn't need to introduce the concept of "stateful" comments either then.

Reply:
I can do this, but in some cases it will result in PRs having several bot comments that may make the PR webpages harder to read. Any time a label or the PR is added/removed/changed a new comment will be added. This will be especially impactful when we initially implement the changelog check. Is the code simplification worth the increase in comments?